src/Conversion/ONNXToKrnl/Tensor/GatherND.cpp

```diff
@@ -64,68 +64,61 @@ struct ONNXGatherNDOpLowering : public OpConversionPattern<ONNXGatherNDOp> {
+    DimsExpr dataDims, indicesDims;
+    create.krnlIE.getShapeAsDims(data, dataDims);
+    create.krnlIE.getShapeAsDims(indices, indicesDims);
     auto dataType = data.getType().cast<ShapedType>();
-    ArrayRef<int64_t> indicesShape = indicesType.getShape();
     ArrayRef<int64_t> dataShape = dataType.getShape();
-    int64_t dataRank = dataShape.size();
-    int64_t indicesRank = indicesShape.size();
+    int64_t dataRank = dataDims.size();
+    int64_t indicesRank = indicesDims.size();
+    auto indicesType = indices.getType().cast<ShapedType>();
+    ArrayRef<int64_t> indicesShape = indicesType.getShape();
     int64_t indicesLastDim = indicesShape[indicesRank - 1];
+    assert((indicesLastDim >= 1 && indicesLastDim <= dataRank - b) &&
+           "indices.shape[-1] must be in the range [1, dataRank - b]");
 
     // Convert the output type to MemRefType.
     Type convertedType = typeConverter->convertType(*op->result_type_begin());
     assert(convertedType && convertedType.isa<MemRefType>() &&
            "Failed to convert type to MemRefType");
-    MemRefType outputMemRefType = convertedType.cast<MemRefType>();
-    ArrayRef<int64_t> outputShape = outputMemRefType.getShape();
-    int64_t outputRank = outputShape.size();
-
-    // Ensure the operation constains are satisfied.
-    assert(dataRank >= 1 && "The rank of 'data' must be >= 1");
-    assert(indicesRank >= 1 && "The rank of 'indices' must be >= 1");
-    assert((outputRank == dataRank + indicesRank - indicesLastDim - 1 - b) &&
-           "Incorrect outut rank");
-    assert(b >= 0 && "batch_dim should not be negative");
-    assert(b < std::min(dataRank, indicesRank) &&
-           "batch_dims must be smaller than the min(dataRank, indicesRank)");
-    assert((indicesLastDim >= 1 && indicesLastDim <= dataRank - b) &&
-           "indices.shape[-1] must be in the range [1, dataRank - b]");
+    DimsExpr outputDims = shapeHelper.getOutputDims();
 
     // Reshape 'indices' to the 3D shape:
     //   [batchDimSize, indicesDimsSize, indices.shape[-1]].
-    const int64_t batchDimsSize = std::accumulate(indicesShape.begin(),
-        indicesShape.begin() + b, 1, std::multiplies<int64_t>());
-    const int64_t indicesDimsSize = std::accumulate(indicesShape.begin(),
-        indicesShape.end(), 1, std::multiplies<int64_t>());
-    assert(batchDimsSize >= 0 && "batchDimsSize must be non-negative");
-    assert(indicesDimsSize >= 0 && "indicesDimsSize must be non-negative");
-
-    LiteralIndexExpr BDS(batchDimsSize),
-        IDS(indicesDimsSize / (batchDimsSize * indicesLastDim)),
-        ILD(indicesLastDim);
+    LiteralIndexExpr oneIE(1);
+    IndexExpr batchDimsSize = oneIE;
+    for (int64_t i = 0; i < b; i++)
+      batchDimsSize = batchDimsSize * indicesDims[i];
+    IndexExpr indicesDimsSize = oneIE;
+    for (int64_t i = 0; i < indicesRank; i++)
+      indicesDimsSize = indicesDimsSize * indicesDims[i];
+    IndexExpr BDS(batchDimsSize),
```

Comment:
Quick question: is this reshape valid even when `indicesDimSizes` is not divisible by `batchDimSize * indicesLastDim` ? Or is it guaranteed? Maybe you can write a comment explaining why this works here in the code.

Reply:
@AlexandreEichenberger Thanks for the comments.

Alex> Quick question: is this reshape valid even when `indicesDimSizes` is not divisible by `batchDimSize * indicesLastDim` ?
Alex> Or is it guaranteed? Maybe you can write a comment explaining why this works here in the code.

It is guaranteed, because IDS can be calculated by product of a part of the indices dimensions.
I changed the code to calculate IDS by using "product" instead of "floorDiv" to clarify the meaning.